### PR TITLE
Add reverb with fast RIR generator

### PR DESCRIPTION
**Pull request by desh2608**
This PR adds the functionality to synthesize a room impulse response on-the-fly for using the `reverb_rir()` methods in the `Cut` or `Recording` (see discussion in https://github.com/lhotse-speech/lhotse/discussions/787). The algorithm is based on [this paper](https://arxiv.org/abs/2208.04101). Here is an audio sample: [Google drive link](https://drive.google.com/file/d/1RGF5rIc66lkWVye100Mi1mhu-lzzxwkT/view?usp=sharing).

**Usage:** 
Basically, all the same methods would still work, but will use the random RIR generator if no RIR recording is explicitly provided.

### CODE REVIEW

File: lhotse/augmentation/utils.py [hunk 1]
```diff
@@ -1,4 +1,9 @@
+from typing import List
+
+import numpy as np
 import torch
+from torchaudio.functional import highpass_biquad
+from torchaudio.transforms import Resample
```
Comment (pzelasko) [inline lhotse/augmentation/utils.py:6]: Can you remove torchaudio imports from the global namespace into the local function you added? They tend to crash documentation builds last time I checked, because they tried to load some shared library that is not available in that env.

Also, I suggest importing `get_or_create_resampler` from `lhotse.augmentation.torchaudio` (locally in the function) as it will create a global resampler, caching the filter coefficients (I got a big speedup from this previously).

File: lhotse/augmentation/utils.py [hunk 2]
```diff
@@ -69,3 +74,134 @@ def convolve1d(signal: torch.Tensor, kernel: torch.Tensor) -> torch.Tensor:
     result = irfft(f_result, n=fast_ftt_size)
 
     return result[:padded_size]
+
+
+# The following is taken from: https://github.com/yluo42/FRA-RIR/blob/main/FRA-RIR.py
+def generate_fast_random_rir(
```
Comment (pzelasko) [inline lhotse/augmentation/utils.py:80]: If it turns out to be too slow, I would check if it's faster with pure `numpy` that with CPU `torch`, that often turned out to be true for me. If it's too much hassle it's OK to leave as is.

File: lhotse/augmentation/utils.py (hunk 1 above)
Comment (desh2608) [reply lhotse/augmentation/utils.py:6]: Done. Thanks for the tip!

**Comment by desh2608** [comment]
Some time comparisons:

1. Reverb with RIR specified
```python
In [14]: %timeit reverb_cut_rir(c,r)
The slowest run took 8.14 times longer than the fastest. This could mean that an intermediate result is being cached.
148 ms ± 88.3 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

2. Reverb with generated RIR (torchaudio `Resample`)
```python
In [9]: %timeit reverb_cut(c)
179 ms ± 57.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

3. Reverb with generated RIR (cached `Resample`)
```python
In [15]: %timeit reverb_cut(c)
138 ms ± 59.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

4. Using Numpy instead of torch
```python
In [6]: %timeit reverb_cut(c)
281 ms ± 121 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Based on above, I'm keeping the torch operations with cached resamplers.

**Comment by pzelasko** [comment]
Wow, nice speed up on PyTorch side. Let’s fix the style checks and LGTM!

**Comment by pzelasko** [comment]
Desh do you have any insight how much simulated vs real RIR matters for tasks such as ASR or diarization? I wonder if given this implementation it still makes sense to use RIR datasets. The audio sample you uploaded sounds nice actually.

**Comment by desh2608** [comment]
> Desh do you have any insight how much simulated vs real RIR matters for tasks such as ASR or diarization? I wonder if given this implementation it still makes sense to use RIR datasets. The audio sample you uploaded sounds nice actually.

I haven't run comparisons myself yet, but at least in the paper it seems to match the results using real RIRs on speech enhancement tasks. Also, older studies have found that at least for ASR, simulated vs real RIRs do not make much difference, so I would imagine that this method should be comparable with those. 

In any case, I am hoping to prepare an AMI recipe for icefall soon, and I can conduct such a comparison then.

**Comment by pzelasko** [comment]
Thanks!

File: lhotse/augmentation/utils.py (hunk 1 above)
Comment (mthrok) [reply lhotse/augmentation/utils.py]: > Can you remove torchaudio imports from the global namespace into the local function you added? They tend to crash documentation builds last time I checked, because they tried to load some shared library that is not available in that env.

@pzelasko Do you have a log of the crash? We make sure that PyTorch is the only hard requirement. I am curious to see what it is.

Comment (pzelasko) [reply lhotse/augmentation/utils.py]: Sure, let me check it again in a spare moment. I think it was related to loading libsndfile.so which was not present in that system. It was also a while ago so maybe the current versions don't have that issue.
